File: internal/services/sonarr/sonarr.go
// Package sonarr provides helpers for the Sonarr v3 API.
package sonarr

import (
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"strings"

	"github.com/rinseaid/reclaimer/internal/services/httpclient"
)

// apiReq performs an HTTP request against the Sonarr API.
func apiReq(sonarrURL, sonarrKey, method, path string, body any, params map[string]string) (*http.Response, error) {
	url := strings.TrimRight(sonarrURL, "/") + "/api/v3" + path

	headers := map[string]string{
		"X-Api-Key": sonarrKey,
	}

	var bodyReader io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("marshal body: %w", err)
		}
		bodyReader = strings.NewReader(string(b))
		headers["Content-Type"] = "application/json"
	}

	return httpclient.Do(httpclient.Request{
		Method:  method,
		URL:     url,
		Headers: headers,
		Params:  params,
		Body:    bodyReader,
	})
}

// apiJSON performs a request and decodes the JSON response into dst.
func apiJSON(sonarrURL, sonarrKey, method, path string, body any, params map[string]string, dst any) error {
	resp, err := apiReq(sonarrURL, sonarrKey, method, path, body, params)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		b, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("sonarr %s %s: HTTP %d: %s", method, path, resp.StatusCode, string(b))
	}

	if dst != nil {
		return json.NewDecoder(resp.Body).Decode(dst)
	}
	return nil
}

// FetchShows returns {tvdbId -> show} with a "_tag_names" list attached to each show.
func FetchShows(sonarrURL, sonarrKey string) (map[int]map[string]any, error) {
	var shows []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/series", nil, nil, &shows); err != nil {
		return nil, fmt.Errorf("fetch shows: %w", err)
	}

	var tags []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/tag", nil, nil, &tags); err != nil {
		return nil, fmt.Errorf("fetch tags: %w", err)
	}

	tagMap := make(map[int]string, len(tags))
	for _, t := range tags {
		id := toInt(t["id"])
		label, _ := t["label"].(string)
		tagMap[id] = label
	}

	result := make(map[int]map[string]any, len(shows))
	for _, s := range shows {
		tvdbID := toInt(s["tvdbId"])
		tagIDs := toIntSlice(s["tags"])
		tagNames := make([]string, 0, len(tagIDs))
		for _, tid := range tagIDs {
			if name, ok := tagMap[tid]; ok {
				tagNames = append(tagNames, name)
			} else {
				tagNames = append(tagNames, strconv.Itoa(tid))
			}
		}
		s["_tag_names"] = tagNames
		result[tvdbID] = s
	}

	slog.Info("Sonarr: fetched shows", "count", len(result))
	return result, nil
}

// FetchShowByTvdb fetches a single show by TVDB ID. Returns {tvdbId -> show} with _tag_names.
func FetchShowByTvdb(sonarrURL, sonarrKey string, tvdbID int) (map[int]map[string]any, error) {
	if sonarrURL == "" || sonarrKey == "" || tvdbID == 0 {
		return map[int]map[string]any{}, nil
	}

	var shows []map[string]any
	err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/series", nil,
		map[string]string{"tvdbId": strconv.Itoa(tvdbID)}, &shows)
	if err != nil {
		return map[int]map[string]any{}, nil
	}
	if len(shows) == 0 {
		return map[int]map[string]any{}, nil
	}

	var tags []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/tag", nil, nil, &tags); err != nil {
		return map[int]map[string]any{}, nil
	}

	tagMap := make(map[int]string, len(tags))
	for _, t := range tags {
		tagMap[toInt(t["id"])], _ = t["label"].(string)
	}

	result := make(map[int]map[string]any, len(shows))
	for _, s := range shows {
		tid := toInt(s["tvdbId"])
		tagIDs := toIntSlice(s["tags"])
		tagNames := make([]string, 0, len(tagIDs))
		for _, id := range tagIDs {
			if name, ok := tagMap[id]; ok {
				tagNames = append(tagNames, name)
			} else {
				tagNames = append(tagNames, strconv.Itoa(id))
			}
		}
		s["_tag_names"] = tagNames
		result[tid] = s
	}
	return result, nil
}

// Unmonitor sets monitored=false for a series.
func Unmonitor(sonarrURL, sonarrKey string, seriesID int, title string) error {
	var show map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, fmt.Sprintf("/series/%d", seriesID), nil, nil, &show); err != nil {
		return fmt.Errorf("unmonitor get series %d: %w", seriesID, err)
	}
	show["monitored"] = false
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPut, fmt.Sprintf("/series/%d", seriesID), show, nil, nil); err != nil {
		return fmt.Errorf("unmonitor put series %d: %w", seriesID, err)
	}
	slog.Info("Unmonitored in Sonarr", "title", title)
	return nil
}

// Delete removes a series from Sonarr.
func Delete(sonarrURL, sonarrKey string, seriesID int, title string, deleteFiles bool) error {
	params := map[string]string{
		"deleteFiles": strconv.FormatBool(deleteFiles),
	}
	resp, err := apiReq(sonarrURL, sonarrKey, http.MethodDelete, fmt.Sprintf("/series/%d", seriesID), nil, params)
	if err != nil {
		return fmt.Errorf("delete series %d: %w", seriesID, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		b, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("delete series %d: HTTP %d: %s", seriesID, resp.StatusCode, string(b))
	}
	slog.Info("Deleted from Sonarr", "title", title, "deleteFiles", deleteFiles)
	return nil
}

// UnmonitorSeason sets monitored=false for a single season without affecting others.
func UnmonitorSeason(sonarrURL, sonarrKey string, seriesID, seasonNumber int, title string) error {
	var show map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, fmt.Sprintf("/series/%d", seriesID), nil, nil, &show); err != nil {
		return fmt.Errorf("unmonitor season get series %d: %w", seriesID, err)
	}

	seasons, _ := show["seasons"].([]any)
	for _, s := range seasons {
		season, ok := s.(map[string]any)
		if !ok {
			continue
		}
		if toInt(season["seasonNumber"]) == seasonNumber {
			season["monitored"] = false
			break
		}
	}

	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPut, fmt.Sprintf("/series/%d", seriesID), show, nil, nil); err != nil {
		return fmt.Errorf("unmonitor season put series %d: %w", seriesID, err)
	}
	slog.Info("Unmonitored season in Sonarr", "season", seasonNumber, "title", title)
	return nil
}

// DeleteSeasonFiles deletes all episode files for a specific season.
func DeleteSeasonFiles(sonarrURL, sonarrKey string, seriesID, seasonNumber int, title string) error {
	var files []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/episodefile", nil,
		map[string]string{"seriesId": strconv.Itoa(seriesID)}, &files); err != nil {
		return fmt.Errorf("delete season files fetch %d: %w", seriesID, err)
	}

	var seasonFiles []map[string]any
	for _, f := range files {
		if toInt(f["seasonNumber"]) == seasonNumber {
			seasonFiles = append(seasonFiles, f)
		}
	}

	deleted := 0
	for _, ef := range seasonFiles {
		fileID := toInt(ef["id"])
		resp, err := apiReq(sonarrURL, sonarrKey, http.MethodDelete, fmt.Sprintf("/episodefile/%d", fileID), nil, nil)
		if err != nil {
			slog.Warn("Failed to delete episode file", "fileId", fileID, "title", title, "season", seasonNumber, "error", err)
			continue
		}
		resp.Body.Close()
		if resp.StatusCode >= 400 {
			slog.Warn("Failed to delete episode file", "fileId", fileID, "title", title, "season", seasonNumber, "status", resp.StatusCode)
			continue
		}
		deleted++
	}

	slog.Info("Deleted season files", "deleted", deleted, "total", len(seasonFiles), "title", title, "season", seasonNumber)
	return nil
}

// SearchSeason triggers a SeasonSearch command.
func SearchSeason(sonarrURL, sonarrKey string, seriesID, seasonNumber int) error {
	body := map[string]any{
		"name":         "SeasonSearch",
		"seriesId":     seriesID,
		"seasonNumber": seasonNumber,
	}
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPost, "/command", body, nil, nil); err != nil {
		return fmt.Errorf("search season: %w", err)
	}
	return nil
}

// SearchSeries triggers a SeriesSearch command.
func SearchSeries(sonarrURL, sonarrKey string, seriesID int) error {
	body := map[string]any{
		"name":     "SeriesSearch",
		"seriesId": seriesID,
	}
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPost, "/command", body, nil, nil); err != nil {
		return fmt.Errorf("search series: %w", err)
	}
	return nil
}

// EnsureTagID returns the ID of a tag with the given label, creating it if needed.
func EnsureTagID(sonarrURL, sonarrKey, label string) (int, error) {
	label = strings.TrimSpace(label)
	if label == "" {
		return 0, fmt.Errorf("empty tag label")
	}

	var tags []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/tag", nil, nil, &tags); err != nil {
		return 0, fmt.Errorf("fetch tags: %w", err)
	}

	for _, t := range tags {
		tLabel, _ := t["label"].(string)
		if strings.EqualFold(strings.TrimSpace(tLabel), label) {
			return toInt(t["id"]), nil
		}
	}

	var created map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPost, "/tag", map[string]string{"label": label}, nil, &created); err != nil {
		return 0, fmt.Errorf("create tag %q: %w", label, err)
	}
	return toInt(created["id"]), nil
}

// AddTag adds a tag to a series by label, creating the tag if missing. Idempotent.
func AddTag(sonarrURL, sonarrKey string, seriesID int, label, title string) error {
	tagID, err := EnsureTagID(sonarrURL, sonarrKey, label)
	if err != nil {
		return err
	}

	var show map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, fmt.Sprintf("/series/%d", seriesID), nil, nil, &show); err != nil {
		return fmt.Errorf("add tag get series %d: %w", seriesID, err)
	}

	tagIDs := toIntSlice(show["tags"])
	for _, id := range tagIDs {
		if id == tagID {
			return nil // already present
		}
	}

	tagIDs = append(tagIDs, tagID)
	show["tags"] = tagIDs
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPut, fmt.Sprintf("/series/%d", seriesID), show, nil, nil); err != nil {
		return fmt.Errorf("add tag put series %d: %w", seriesID, err)
	}
	slog.Info("Added Sonarr tag", "label", label, "title", title)
	return nil
}

// RemoveTag removes a tag from a series by label. No-op if not present.
func RemoveTag(sonarrURL, sonarrKey string, seriesID int, label, title string) error {
	label = strings.TrimSpace(label)
	if label == "" {
		return nil
	}

	var tags []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/tag", nil, nil, &tags); err != nil {
		return fmt.Errorf("remove tag fetch tags: %w", err)
	}

	wantID := -1
	for _, t := range tags {
		tLabel, _ := t["label"].(string)
		if strings.EqualFold(strings.TrimSpace(tLabel), label) {
			wantID = toInt(t["id"])
			break
		}
	}
	if wantID < 0 {
		return nil
	}

	var show map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, fmt.Sprintf("/series/%d", seriesID), nil, nil, &show); err != nil {
		return fmt.Errorf("remove tag get series %d: %w", seriesID, err)
	}

	before := toIntSlice(show["tags"])
	after := make([]int, 0, len(before))
	for _, id := range before {
		if id != wantID {
			after = append(after, id)
		}
	}
	if len(after) == len(before) {
		return nil
	}

	show["tags"] = after
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPut, fmt.Sprintf("/series/%d", seriesID), show, nil, nil); err != nil {
		return fmt.Errorf("remove tag put series %d: %w", seriesID, err)
	}
	slog.Info("Removed Sonarr tag", "label", label, "title", title)
	return nil
}

// SetRootFolder moves a series to a different root folder via /series/editor.
func SetRootFolder(sonarrURL, sonarrKey string, seriesID int, newRoot string, moveFiles bool, title string) error {
	newRoot = strings.TrimSpace(newRoot)
	if newRoot == "" {
		return fmt.Errorf("root folder path is required")
	}

	body := map[string]any{
		"seriesIds":      []int{seriesID},
		"rootFolderPath": newRoot,
		"moveFiles":      moveFiles,
	}
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPut, "/series/editor", body, nil, nil); err != nil {
		return fmt.Errorf("set root folder series %d: %w", seriesID, err)
	}
	slog.Info("Moved Sonarr series", "title", title, "root", newRoot, "moveFiles", moveFiles)
	return nil
}

// AddSeries looks up a series by TVDB ID and adds it to Sonarr.
func AddSeries(sonarrURL, sonarrKey string, tvdbID int, title string, qualityProfileID int, rootFolder string, monitored, searchOnAdd bool, tags []int) (map[string]any, error) {
	var lookup []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/series/lookup", nil,
		map[string]string{"term": fmt.Sprintf("tvdb:%d", tvdbID)}, &lookup); err != nil {
		return nil, fmt.Errorf("series lookup tvdb %d: %w", tvdbID, err)
	}
	if len(lookup) == 0 {
		return nil, fmt.Errorf("TVDB id %d not found on Sonarr", tvdbID)
	}

	payload := lookup[0]
	payload["qualityProfileId"] = qualityProfileID
	payload["rootFolderPath"] = rootFolder
	payload["monitored"] = monitored
	if _, ok := payload["seasonFolder"]; !ok {
		payload["seasonFolder"] = true
	}
	if tags == nil {
		tags = []int{}
	}
	payload["tags"] = tags
	payload["addOptions"] = map[string]any{
		"searchForMissingEpisodes":       searchOnAdd,
		"searchForCutoffUnmetEpisodes":   false,
		"monitor":                        "all",
	}

	var created map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodPost, "/series", payload, nil, &created); err != nil {
		return nil, fmt.Errorf("add series tvdb %d: %w", tvdbID, err)
	}
	slog.Info("Added series to Sonarr", "title", title, "tvdbId", tvdbID, "id", created["id"])
	return created, nil
}

// ListQualityProfiles returns [{id, name}, ...] from /qualityprofile.
func ListQualityProfiles(sonarrURL, sonarrKey string) ([]map[string]any, error) {
	if sonarrURL == "" || sonarrKey == "" {
		return nil, nil
	}

	var profiles []map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/qualityprofile", nil, nil, &profiles); err != nil {
		slog.Warn("Sonarr list_quality_profiles failed", "error", err)
		return nil, nil
	}

	result := make([]map[string]any, 0, len(profiles))
	for _, p := range profiles {
		result = append(result, map[string]any{
			"id":   p["id"],
			"name": p["name"],
		})
	}
	return result, nil
}

// GetQualityProfileID finds a quality profile ID by name (case insensitive).
// Returns the ID and true if found, or 0 and false otherwise.
func GetQualityProfileID(sonarrURL, sonarrKey, name string) (int, bool) {
	if name == "" {
		return 0, false
	}
	target := strings.TrimSpace(strings.ToLower(name))
	profiles, _ := ListQualityProfiles(sonarrURL, sonarrKey)
	for _, p := range profiles {
		pName, _ := p["name"].(string)
		if strings.TrimSpace(strings.ToLower(pName)) == target {
			id := toInt(p["id"])
			if id != 0 {
				return id, true
			}
		}
	}
	return 0, false
}

// RecycleBinPath returns Sonarr's configured recycle bin path, or empty string on failure.
func RecycleBinPath(sonarrURL, sonarrKey string) string {
	if sonarrURL == "" || sonarrKey == "" {
		return ""
	}

	var config map[string]any
	if err := apiJSON(sonarrURL, sonarrKey, http.MethodGet, "/config/mediamanagement", nil, nil, &config); err != nil {
		slog.Warn("Sonarr recycle_bin_path lookup failed", "error", err)
		return ""
	}

	path, _ := config["recycleBin"].(string)
	return strings.TrimSpace(path)
}

// BuildSeasonCounts computes {series_id -> highest seasonNumber with files}. Skips season 0.
func BuildSeasonCounts(shows map[int]map[string]any) map[int]int {
	counts := make(map[int]int)
	for _, show := range shows {
		seriesID := toInt(show["id"])
		if seriesID == 0 {
			continue
		}

		highest := -1
		seasons, _ := show["seasons"].([]any)
		for _, s := range seasons {
			season, ok := s.(map[string]any)
			if !ok {
				continue
			}
			seasonNum := toInt(season["seasonNumber"])
			if seasonNum <= 0 {
				continue
			}

			stats, _ := season["statistics"].(map[string]any)
			if stats == nil {
				continue
			}
			fileCount := toInt(stats["episodeFileCount"])
			if fileCount > 0 && seasonNum > highest {
				highest = seasonNum
			}
		}

		if highest > 0 {
			counts[seriesID] = highest
		}
	}
	return counts
}

// toInt converts a JSON number (float64) or other numeric value to int.
func toInt(v any) int {
	switch n := v.(type) {
	case float64:
		return int(n)
	case int:
		return n
	case int64:
		return int(n)
	case json.Number:
		i, _ := n.Int64()
		return int(i)
	default:
		return 0
	}
}

// toIntSlice converts a JSON array of numbers to []int.
func toIntSlice(v any) []int {
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	result := make([]int, 0, len(arr))
	for _, item := range arr {
		result = append(result, toInt(item))
	}
	return result
}

File: app/core/sonarr.py
"""Sonarr API helpers."""
from __future__ import annotations

import logging
from .clients import get_client

log = logging.getLogger(__name__)


def _api(sonarr_url: str, sonarr_key: str, method: str, path: str, **kwargs):
    return getattr(get_client(), method)(
        f"{sonarr_url}/api/v3{path}",
        headers={"X-Api-Key": sonarr_key},
        timeout=30,
        **kwargs,
    )


def fetch_shows(sonarr_url: str, sonarr_key: str) -> dict[int, dict]:
    """Return {tvdbId → show_dict} with _tag_names list attached."""
    shows = _api(sonarr_url, sonarr_key, "get", "/series").json()
    tag_map = {t["id"]: t["label"] for t in _api(sonarr_url, sonarr_key, "get", "/tag").json()}
    result: dict[int, dict] = {}
    for s in shows:
        s["_tag_names"] = [tag_map.get(t, str(t)) for t in s.get("tags", [])]
        result[s["tvdbId"]] = s
    log.info("Sonarr: %d shows", len(result))
    return result


def unmonitor(sonarr_url: str, sonarr_key: str, sonarr_id: int, title: str) -> None:
    r = _api(sonarr_url, sonarr_key, "get", f"/series/{sonarr_id}")
    r.raise_for_status()
    show = r.json()
    show["monitored"] = False
    _api(sonarr_url, sonarr_key, "put", f"/series/{sonarr_id}", json=show).raise_for_status()
    log.info("Unmonitored in Sonarr: %s", title)


def delete(sonarr_url: str, sonarr_key: str, sonarr_id: int, title: str,
           delete_files: bool = True) -> None:
    _api(sonarr_url, sonarr_key, "delete", f"/series/{sonarr_id}",
         params={"deleteFiles": str(delete_files).lower()}).raise_for_status()
    log.info("Deleted from Sonarr: %s (deleteFiles=%s)", title, delete_files)


def fetch_show_by_tvdb(sonarr_url: str, sonarr_key: str, tvdb_id: int) -> dict[int, dict]:
    """Fetch a single show by TVDB ID. Returns {tvdbId -> show_dict} with _tag_names."""
    if not sonarr_url or not sonarr_key or not tvdb_id:
        return {}
    try:
        shows = _api(sonarr_url, sonarr_key, "get", "/series", params={"tvdbId": tvdb_id}).json()
        if not shows:
            return {}
        tag_map = {t["id"]: t["label"] for t in _api(sonarr_url, sonarr_key, "get", "/tag").json()}
        result: dict[int, dict] = {}
        for s in shows:
            s["_tag_names"] = [tag_map.get(t, str(t)) for t in s.get("tags", [])]
            result[s["tvdbId"]] = s
        return result
    except Exception:
        return {}


def unmonitor_season(sonarr_url: str, sonarr_key: str, series_id: int,
                     season_number: int, title: str) -> None:
    """Unmonitor a single season without affecting other seasons."""
    r = _api(sonarr_url, sonarr_key, "get", f"/series/{series_id}")
    r.raise_for_status()
    show = r.json()
    for s in show.get("seasons", []):
        if s.get("seasonNumber") == season_number:
            s["monitored"] = False
            break
    _api(sonarr_url, sonarr_key, "put", f"/series/{series_id}", json=show).raise_for_status()
    log.info("Unmonitored season %d in Sonarr: %s", season_number, title)


def delete_season_files(sonarr_url: str, sonarr_key: str, series_id: int,
                        season_number: int, title: str) -> None:
    """Delete all episode files for a specific season."""
    r = _api(sonarr_url, sonarr_key, "get", "/episodefile",
             params={"seriesId": series_id})
    r.raise_for_status()
    files = r.json()
    season_files = [f for f in files if f.get("seasonNumber") == season_number]
    deleted = 0
    for ef in season_files:
        try:
            _api(sonarr_url, sonarr_key, "delete",
                 f"/episodefile/{ef['id']}").raise_for_status()
            deleted += 1
        except Exception as e:
            log.warning("Failed to delete episode file %d for %s S%02d: %s",
                        ef.get("id", 0), title, season_number, e)
    log.info("Deleted %d/%d files for %s season %d",
             deleted, len(season_files), title, season_number)


def search_season(sonarr_url: str, sonarr_key: str, series_id: int, season_number: int) -> None:
    """Trigger a season search."""
    _api(sonarr_url, sonarr_key, "post", "/command",
         json={"name": "SeasonSearch", "seriesId": series_id,
               "seasonNumber": season_number}).raise_for_status()


def search_series(sonarr_url: str, sonarr_key: str, series_id: int) -> None:
    """Trigger a full-series search."""
    _api(sonarr_url, sonarr_key, "post", "/command",
         json={"name": "SeriesSearch", "seriesId": int(series_id)}).raise_for_status()


def ensure_tag_id(url: str, key: str, label: str) -> int | None:
    """Return the id of the Sonarr tag with ``label``, creating it if
    needed. Mirrors ``radarr.ensure_tag_id``."""
    label = (label or "").strip()
    if not label:
        return None
    r = _api(url, key, "get", "/tag")
    r.raise_for_status()
    for t in r.json():
        if (t.get("label") or "").strip().lower() == label.lower():
            return int(t["id"])
    r2 = _api(url, key, "post", "/tag", json={"label": label})
    r2.raise_for_status()
    return int(r2.json()["id"])


def add_tag(url: str, key: str, series_id: int, label: str, title: str = "") -> None:
    """Add a tag to a series by label, creating the tag if missing. No-op
    when the tag is already present (idempotent)."""
    tag_id = ensure_tag_id(url, key, label)
    if tag_id is None:
        return
    r = _api(url, key, "get", f"/series/{series_id}")
    r.raise_for_status()
    show = r.json()
    tags = list(show.get("tags") or [])
    if tag_id in tags:
        return
    tags.append(tag_id)
    show["tags"] = tags
    _api(url, key, "put", f"/series/{series_id}", json=show).raise_for_status()
    log.info("Added Sonarr tag %r to %s", label, title or series_id)


def remove_tag(url: str, key: str, series_id: int, label: str, title: str = "") -> None:
    """Remove a tag from a series by label."""
    label = (label or "").strip()
    if not label:
        return
    tags_r = _api(url, key, "get", "/tag")
    tags_r.raise_for_status()
    want = None
    for t in tags_r.json():
        if (t.get("label") or "").strip().lower() == label.lower():
            want = int(t["id"])
            break
    if want is None:
        return
    r = _api(url, key, "get", f"/series/{series_id}")
    r.raise_for_status()
    show = r.json()
    before = list(show.get("tags") or [])
    after = [t for t in before if int(t) != want]
    if len(after) == len(before):
        return
    show["tags"] = after
    _api(url, key, "put", f"/series/{series_id}", json=show).raise_for_status()
    log.info("Removed Sonarr tag %r from %s", label, title or series_id)


def set_root_folder(url: str, key: str, series_id: int, new_root: str,
                    move_files: bool = True, title: str = "") -> None:
    """Move a series to a different root folder via Sonarr's bulk-edit
    endpoint (``/series/editor``). Sonarr performs the file move itself."""
    new_root = (new_root or "").strip()
    if not new_root:
        raise ValueError("root folder path is required")
    body = {
        "seriesIds": [int(series_id)],
        "rootFolderPath": new_root,
        "moveFiles": bool(move_files),
    }
    r = _api(url, key, "put", "/series/editor", json=body)
    r.raise_for_status()
    log.info("Moved Sonarr series %s to root %s (moveFiles=%s)",
             title or series_id, new_root, move_files)


def add_series(url: str, key: str, tvdb_id: int, title: str,
               quality_profile_id: int, root_folder: str,
               monitored: bool = True, search_on_add: bool = False,
               tags: list[int] | None = None) -> dict:
    """Create a series on this Sonarr via ``/series/lookup?term=tvdb:ID``.
    Returns the created series dict."""
    r = _api(url, key, "get", "/series/lookup",
             params={"term": f"tvdb:{int(tvdb_id)}"})
    r.raise_for_status()
    lookup = r.json() or []
    if not lookup:
        raise ValueError(f"TVDB id {tvdb_id} not found on Sonarr")
    payload = dict(lookup[0])
    payload["qualityProfileId"] = int(quality_profile_id)
    payload["rootFolderPath"] = root_folder
    payload["monitored"] = bool(monitored)
    payload["seasonFolder"] = payload.get("seasonFolder", True)
    payload["tags"] = list(tags or [])
    payload["addOptions"] = {
        "searchForMissingEpisodes": bool(search_on_add),
        "searchForCutoffUnmetEpisodes": False,
        "monitor": "all",
    }
    cr = _api(url, key, "post", "/series", json=payload)
    cr.raise_for_status()
    created = cr.json()
    log.info("Added series %r to Sonarr (tvdbId=%s, id=%s)",
             title, tvdb_id, created.get("id"))
    return created


def list_quality_profiles(url: str, key: str) -> list[dict]:
    """Return ``[{id, name}, ...]`` from ``/api/v3/qualityprofile``.

    Wrapped in try/except so orchestration doesn't abort when Sonarr is
    unreachable -- callers can tell "no profiles known" from an empty list.
    """
    if not url or not key:
        return []
    try:
        r = _api(url, key, "get", "/qualityprofile")
        r.raise_for_status()
        return [{"id": p.get("id"), "name": p.get("name")} for p in r.json()]
    except Exception as e:
        log.warning("Sonarr list_quality_profiles failed: %s", e)
        return []


def recycle_bin_path(url: str, key: str) -> str | None:
    """Return Sonarr's Media Management recycle-bin path, or ``None`` when
    it's unset / unreachable. Mirrors radarr.recycle_bin_path; used by the
    rule editor to warn when a delete-files step has no safety net.
    """
    if not url or not key:
        return None
    try:
        r = _api(url, key, "get", "/config/mediamanagement")
        r.raise_for_status()
        path = (r.json() or {}).get("recycleBin") or ""
        return path.strip() or None
    except Exception as e:
        log.warning("Sonarr recycle_bin_path lookup failed: %s", e)
        return None


def get_quality_profile_id(url: str, key: str, name: str) -> int | None:
    """Return the profile ``id`` for the named profile, or ``None`` if not
    found / API unreachable. Name match is case-insensitive.
    """
    if not name:
        return None
    target = name.strip().lower()
    for p in list_quality_profiles(url, key):
        pname = (p.get("name") or "").strip().lower()
        if pname == target and p.get("id") is not None:
            try:
                return int(p["id"])
            except (TypeError, ValueError):
                return None
    return None


def build_season_counts(shows: dict[int, dict]) -> dict[int, int]:
    """Compute {series_id -> highest seasonNumber with any episode files}.

    Skips Season 0 (specials). Shows with no populated seasons are omitted so
    callers can treat absence as "unknown" rather than "zero seasons".
    """
    counts: dict[int, int] = {}
    for show in shows.values():
        series_id = show.get("id")
        if series_id is None:
            continue
        highest = -1
        for s in show.get("seasons", []) or []:
            season_num = s.get("seasonNumber", 0)
            if season_num <= 0:  # skip specials / malformed
                continue
            file_count = (s.get("statistics") or {}).get("episodeFileCount", 0) or 0
            if file_count > 0 and season_num > highest:
                highest = season_num
        if highest > 0:
            counts[int(series_id)] = highest
    return counts
